### ISSUE TITLE
A way to release an object

### ISSUE THREAD

**Issue by Vorlex** [ssannandeji/Zenject-2019#441]
Hi!

Currently rewriting usage of DI container from Uniject to Zenject in our Unity project. We actively use theirs (Uniject) Kernel.Release(object), which allows us to release created instance of a class and use new one with another binding. For example WithId:
`Container.Bind<IPlayerView>().WithId(PlayerViewMode.FirstPersonMode.ToString()).To<FPPlayerView>().AsSingle();`
`Container.Bind<IPlayerView>().WithId(PlayerViewMode.ThirdPersonMode.ToString()).To<TPPlayerView>().AsSingle();`
So whenever we switch to First/Third person view, we need to Release previous IPlayerView.

Does Zenject have some way to do it?

**Comment by sunmachine** [comment]
I think there is a `Rebind()` method, no? Not sure if this does precisely what you need.
https://github.com/modesttree/Zenject#dicontainer-methods-rebind

**Comment by Vorlex** [comment]
Actually I need a way to say Container that I need to release an instantiated instance of an object without knowing anything about this object's bindings.

For example I switched to `ThirdPersonMode` from `FirstPersonMode`.
I need Container to delete info about instantiated `FPPlayerView` (because in our code we release all resources: game objects, camera, etc, that were instantiated in it, and can't use this instance again), and give me `TPPlayerView` when I request `IPlayerView`.
If I switch to `FirstPersonMode` again. Container releases `TPPlayerView ` and so on.

Is there any way to do something like this:
```
IPlayerView playerView = Container.ResolveId<IPlayerView>(PlayerViewMode.FirstPersonMode.ToString())
Container.Release(playerView );
```

**Comment by svermeulen** [comment]
Can you use AsTransient instead?  If you just call Resolve in that one place then zenject won't store a reference to it.

There is the Unbind method, but all that does is remove the 'contract' part of the binding, so the next time somebody requests IPlayerView, then it will throw an exception.  It sounds like you want a method that clears the reference that zenject stores internally to TPPlayerView / FPPlayerView when using AsSingle or AsCached.  Is that right?

**Comment by Vorlex** [comment]
> Can you use AsTransient instead? If you just call Resolve in that one place then zenject won't store a reference to it.

As for IPlayerView indeed we can use AsTransient, but we have a lot of other stuff that is used in more than one place, so we need AsSingle.
For example, IEnvironmentController that is responsible for weather (rain, fog, etc). A lot of classes uses it.

> It sounds like you want a method that clears the reference that zenject stores internally to TPPlayerView / FPPlayerView when using AsSingle or AsCached. Is that right?

Yes, exactly!

**Comment by svermeulen** [comment]
It's possible that we could add a feature like this but I'm not convinced it would be a good idea.

When following DI practices properly, the container should not change after the install phase completes and the first object is instantiated.  At that point you can still technically call Container.Bind but it's dangerous because you might make a change to the container that would have affected previously instantiated objects.

When you do things like call Resolve directly at runtime instead of getting your instances injected, and call Bind to add instances at runtime, etc. you're getting away from using DI and moving towards Service Locator pattern.

I know you aren't talking about using Container.Bind after startup, and are just talking about clearing part of the cache of instances stored in DiContainer, but it's kind of similar, because we would be breaking the semantics of AsSingle.  Since multiple objects could have distinct instances of the same binding that was marked AsSingle.

Can you use a manager class instead that uses factories to create instances of IPlayerView?  And then it could just clear its own reference to ensure that no data carries over

**Comment by Vorlex** [comment]
I agree about practices. But in our [game](https://play.google.com/store/apps/details?id=com.plabs.planetofc&hl=en) we have multiple game modes with a lot of objects that need to be created when launching mode and then released after exiting to main menu to reduce memory usage before launching other mode. We have some sort of cascade structure in which objects are released in recursive manner.
You can check part of our [bindings ](https://github.com/modesttree/Zenject/files/1902564/MainInstaller.txt) to understand the scale of the project.
It is very hard to rewrite current structure to "manager class" technique, because there are a lot more classes like PlayerView.

If you're against adding this feature, could you suggest where I need to look to add it myself.
Or it would be wonderful if you can share some code snippet to help me :)

**Comment by svermeulen** [comment]
Sure I can sympathize with that.  What version are you using?

**Comment by Vorlex** [comment]
Great!
The latest - 5.5.1.

**Comment by svermeulen** [comment]
Ok I added the ability to do this on develop branch which is going to become zenject 6.  With that you can do things like this:

```
var context = new InjectContext(Container, typeof(IFoo));

foreach (var provider in Container.AllProviders.OfType<CachedProvider>()
    .Where(x => x.GetInstanceType(context) == typeof(Foo1)))
{
    provider.ClearCache();
}
```

Which I think can get you what you want.  You might not want to use develop branch so what you could do instead is look at the commit I made above and copy that to your local version.

**Comment by Vorlex** [comment]
Great! Thank you so much!

I will test it and provide a response.

**Comment by Vorlex** [comment]
Tested with this class (actually we don't need it to release, it just came to hand):
`Container.Bind<IRateService>().To<RateService>().AsSingle();`

And got this exception on second resolve:
```
Container.Resolve<IRateService>();
var context = new InjectContext(Container, typeof(IRateService));

foreach (var provider in Container.AllProviders.OfType<CachedProvider>().Where(x => x.GetInstanceType(context) == typeof(RateService)))
{
    provider.ClearCache();
}
//throws exception
Container.Resolve<IRateService>();
```
```
ZenjectException: Provider returned zero instances when one was expected!  While resolving type 'Planet.Service.Rate.IRateService'. 
Object graph:

Zenject.DiContainer.Resolve (Zenject.InjectContext context) (at Assets/Plugins/Zenject/Source/Main/DiContainer.cs:777)
Zenject.DiContainer.ResolveId (System.Type contractType, System.Object identifier) (at Assets/Plugins/Zenject/Source/Main/DiContainer.cs:1828)
Zenject.DiContainer.Resolve (System.Type contractType) (at Assets/Plugins/Zenject/Source/Main/DiContainer.cs:1818)
Zenject.DiContainer.Resolve[IRateService] () (at Assets/Plugins/Zenject/Source/Main/DiContainer.cs:1813)
```

**Comment by svermeulen** [comment]
Oh right, I think `_instances` should be set to null rather then cleared in CachedProvider.  Can you try that?

**Comment by Vorlex** [comment]
Yes, it works!

Thank you again!

**Comment by svermeulen** [comment]
No problem, good luck with your project  :)

**Comment by Vorlex** [comment]
Hi!

I updated to 6.1.1 and now have issue with code you provided me with in this thread:

```
var context = new InjectContext(Container, typeof(IFoo));

foreach (var provider in Container.AllProviders.OfType<CachedProvider>()
    .Where(x => x.GetInstanceType(context) == typeof(Foo1)))
{
    provider.ClearCache();
}
```

Now it throws exception:
```
ZenjectException: Expected type 'TickableManager' to derive from or be equal to 'Foo1'
ModestTree.Assert.DerivesFromOrEqual (System.Type childType, System.Type parentType) (at Assets/Plugins/Zenject/Source/Internal/Assert.cs:104)
Zenject.ProviderUtil.GetTypeToInstantiate (System.Type contractType, System.Type concreteType) (at Assets/Plugins/Zenject/Source/Providers/ProviderUtil.cs:19)
Zenject.TransientProvider.GetTypeToCreate (System.Type contractType) (at Assets/Plugins/Zenject/Source/Providers/TransientProvider.cs:69)
Zenject.TransientProvider.GetInstanceType (Zenject.InjectContext context) (at Assets/Plugins/Zenject/Source/Providers/TransientProvider.cs:41)
Zenject.CachedProvider.GetInstanceType (Zenject.InjectContext context) (at Assets/Plugins/Zenject/Source/Providers/CachedProvider.cs:50)
Planet.Zenjector.ZenjectExtensions+<Release>c__AnonStorey0.<>m__0 (Zenject.CachedProvider x) (at Assets/Plugins/Zenjector/ZenjectExtensions.cs:20)
System.Linq.Enumerable+<CreateWhereIterator>c__Iterator1D`1[Zenject.CachedProvider].MoveNext ()
...
```

I assume it happens because `GetInstanceType()` of `TransientProvider` changed from
`return _concreteType;` to `return GetTypeToCreate(context.MemberType);` which calls `ProviderUtil.GetTypeToInstantiate` which make this assert `Assert.DerivesFromOrEqual(concreteType, contractType)` which throws the exception.

Can you help me with this, please?

**Comment by svermeulen** [comment]
Should work now if you sync from source

**Comment by Vorlex** [comment]
Thanks! It works just fine except for open generic types when I use it like this:
```
IFoo<IBar> fooBar = diContainer.Resolve<IFoo<IBar>>();
var type= diContainer.ResolveType(fooBar.GetType().GetInterfaces()[0].GetGenericTypeDefinition()));
```
It returns null because `TypeExtensions.DerivesFromOrEqual `returns false.
I stumbled upon this on SO https://stackoverflow.com/questions/5461295/using-isassignablefrom-with-open-generic-types and found that `TypeExtensions` already contains that method. Maybe it is a good idea to use `IsAssignableToGenericType` in `DerivesFromOrEqual`?
